lib/internal/errors.js

```diff
@@ -0,0 +1,258 @@
+'use strict';
+
+// The whole point behind this internal module is to allow Node.js to no
+// longer be forced to treat every error message change as a semver-major
+// change. The NodeError classes here all expose a `code` property whose
+// value identifies the error.
+
+const kCode = Symbol('code');
+
+class NodeError extends Error {
+  constructor(key, ...args) {
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeError);
```

Comment:
Could you use `new.target` or `this.constructor` instead of using `NodeError` explicitly? That way it should correctly capture the stack trace relatively to the top most constructor without invoking it again in the child classes.